File: lib/physics.ts
import { PARTS, PlacedPart } from './parts'

export interface PhysicsResult {
  stable: boolean
  score: number
  reasons: string[]
  com: { x: number; y: number; z: number }
  legCount: number
  surfaceCount: number
  perPartStatus: Record<string, 'stable' | 'unstable' | 'neutral'>
}

function getBounds(part: PlacedPart) {
  const def = PARTS[part.type]
  if (!def) return null
  return {
    minX: part.x - def.w / 2, maxX: part.x + def.w / 2,
    minY: part.y - def.h / 2, maxY: part.y + def.h / 2,
    minZ: part.z - def.d / 2, maxZ: part.z + def.d / 2,
    cx: part.x, cy: part.y, cz: part.z,
  }
}

export function validatePhysics(parts: PlacedPart[]): PhysicsResult {
  const reasons: string[] = []
  const perPartStatus: Record<string, 'stable' | 'unstable' | 'neutral'> = {}
  let score = 100
  const surfaces = parts.filter(p => ['surface','cushion'].includes(PARTS[p.type]?.kind))
  const legs = parts.filter(p => PARTS[p.type]?.kind === 'leg')
  const backs = parts.filter(p => PARTS[p.type]?.kind === 'back')
  const panels = parts.filter(p => PARTS[p.type]?.kind === 'panel')
  parts.forEach(p => { perPartStatus[p.id] = 'neutral' })
  if (!parts.length) return { stable: false, score: 0, reasons: ['No parts'], com: {x:0,y:0,z:0}, legCount:0, surfaceCount:0, perPartStatus }
  let tm = 0, cx = 0, cy = 0, cz = 0
  parts.forEach(p => { const d = PARTS[p.type]; if (!d) return; tm += d.kg; cx += p.x*d.kg; cy += p.y*d.kg; cz += p.z*d.kg })
  if (tm > 0) { cx/=tm; cy/=tm; cz/=tm }
  if (!surfaces.length) { reasons.push('No surface'); score -= 40 }
  if (!legs.length && !panels.length) { reasons.push('No support'); score -= 40 }
  for (const surf of surfaces) {
    const sb = getBounds(surf); if (!sb) continue
    const sLow = sb.minY, TOL = 0.12
    const sLegs = legs.filter(l => { const b = getBounds(l); return b && Math.abs(b.maxY - sLow) < TOL })
    const sPanels = panels.filter(p => { const b = getBounds(p); return b && b.maxY >= sLow - TOL })
    if (sLegs.length >= 3 || sPanels.length >= 2) {
      perPartStatus[surf.id] = 'stable'
      sLegs.forEach(l => { perPartStatus[l.id] = 'stable' })
      sPanels.forEach(p => { perPartStatus[p.id] = 'stable' })
      if (sLegs.length > 1) {
        const bots = sLegs.map(l => getBounds(l)!.minY)
        if (Math.max(...bots) - Math.min(...bots) > 0.08) { reasons.push('Unequal leg heights'); score -= 25; perPartStatus[surf.id] = 'unstable'; sLegs.forEach(l => { perPartStatus[l.id] = 'unstable' }) }
      }
      if (sLegs.length >= 2) {
        const lx = sLegs.map(l => l.x), lz = sLegs.map(l => l.z)
        if (cx < Math.min(...lx)-0.05 || cx > Math.max(...lx)+0.05 || cz < Math.min(...lz)-0.05 || cz > Math.max(...lz)+0.05) { reasons.push('CoM outside base'); score -= 30; perPartStatus[surf.id] = 'unstable' }
      }
    } else {
      perPartStatus[surf.id] = 'unstable'
      reasons.push(sLegs.length || sPanels.length ? 'Insufficient support' : 'No support for surface')
      score -= sLegs.length || sPanels.length ? 20 : 35
    }
    backs.forEach(b => {
      const ok = Math.abs(b.x - surf.cx) < (sb.maxX-sb.minX)/2+0.1 && Math.abs(b.z - surf.cz) < 0.3
      perPartStatus[b.id] = ok ? 'stable' : 'unstable'
      if (!ok) { reasons.push('Backrest disconnected'); score -= 15 }
    })
  }
  parts.forEach(p => { if (perPartStatus[p.id] === 'neutral') { const b = getBounds(p); if (b && b.minY < 0.04) perPartStatus[p.id] = 'stable' } })
  score = Math.max(0, Math.min(100, score))
  const stable = score >= 70 && !reasons.length
  if (stable) reasons.push('Structure is stable')
  return { stable, score, reasons, com: {x:cx,y:cy,z:cz}, legCount: legs.length, surfaceCount: surfaces.length, perPartStatus }
}

export function getScoreGrade(score: number): 'A'|'B'|'C'|'D' {
  return score >= 90 ? 'A' : score >= 75 ? 'B' : score >= 50 ? 'C' : 'D'
}

File: backend/services/physics.py
"""
Server-side physics validation engine.
Mirrors the TypeScript logic in lib/physics.ts exactly.
Updated with correct real-product dimensions from lib/parts.ts.
"""
from typing import List, Dict, Any, Optional

# ── Part definitions (mirrors PARTS in lib/parts.ts) ─────────
# All dimensions in meters, kg in kilograms
PARTS: Dict[str, Dict[str, Any]] = {
    # Basic parts
    "seat":           {"kind": "surface", "w": 0.45, "h": 0.04, "d": 0.45, "kg": 2.5},
    "tabletop":       {"kind": "surface", "w": 1.20, "h": 0.04, "d": 0.70, "kg": 8.0},
    "shelf":          {"kind": "surface", "w": 0.80, "h": 0.025,"d": 0.30, "kg": 3.0},
    "leg-short":      {"kind": "leg",     "w": 0.04, "h": 0.45, "d": 0.04, "kg": 0.5},
    "leg-long":       {"kind": "leg",     "w": 0.04, "h": 0.72, "d": 0.04, "kg": 0.8},
    "crossbar":       {"kind": "support", "w": 0.40, "h": 0.03, "d": 0.03, "kg": 0.6},
    "backrest":       {"kind": "back",    "w": 0.44, "h": 0.45, "d": 0.04, "kg": 1.8},
    "armrest":        {"kind": "support", "w": 0.20, "h": 0.03, "d": 0.20, "kg": 0.9},
    "panel":          {"kind": "panel",   "w": 0.04, "h": 1.80, "d": 0.40, "kg": 5.0},
    # Real products — exact dimensions
    "prod-oak-top":     {"kind": "surface", "w": 1.80, "h": 0.04, "d": 0.80, "kg": 22.0},
    "prod-walnut-top":  {"kind": "surface", "w": 1.60, "h": 0.04, "d": 0.75, "kg": 18.0},
    "prod-pine-desk":   {"kind": "surface", "w": 1.60, "h": 0.03, "d": 0.80, "kg": 12.0},
    "prod-bamboo-shelf":{"kind": "surface", "w": 0.80, "h": 0.02, "d": 0.25, "kg": 1.6},
    "prod-dining-seat": {"kind": "surface", "w": 0.42, "h": 0.04, "d": 0.42, "kg": 2.0},
    "prod-cushion":     {"kind": "cushion", "w": 0.40, "h": 0.08, "d": 0.40, "kg": 0.8},
    "prod-hairpin":     {"kind": "leg",     "w": 0.04, "h": 0.71, "d": 0.04, "kg": 0.6},
    "prod-scandi-leg":  {"kind": "leg",     "w": 0.04, "h": 0.43, "d": 0.04, "kg": 0.5},
    "prod-u-leg":       {"kind": "leg",     "w": 0.50, "h": 0.72, "d": 0.04, "kg": 4.2},
    "prod-tall-panel":  {"kind": "panel",   "w": 0.04, "h": 1.80, "d": 0.40, "kg": 8.0},
    "prod-door":        {"kind": "door",    "w": 0.40, "h": 0.70, "d": 0.02, "kg": 3.0},
}

SURFACE_KINDS = frozenset(["surface", "cushion"])


def _is_surface(kind: Optional[str]) -> bool:
    return kind in SURFACE_KINDS


def get_bounds(part: Dict) -> Optional[Dict]:
    """Get bounding box for a placed part."""
    defn = PARTS.get(part.get("type", ""))
    if not defn:
        return None
    x, y, z = part.get("x", 0.0), part.get("y", 0.0), part.get("z", 0.0)
    return {
        "min_x": x - defn["w"] / 2,
        "max_x": x + defn["w"] / 2,
        "min_y": y - defn["h"] / 2,
        "max_y": y + defn["h"] / 2,
        "min_z": z - defn["d"] / 2,
        "max_z": z + defn["d"] / 2,
        "cx": x, "cy": y, "cz": z,
    }


def validate_design(parts: List[Dict]) -> Dict:
    """
    Validate furniture design physics. Mirrors lib/physics.ts:validatePhysics().

    Args:
        parts: List of placed parts [{id, type, x, y, z, ...}]

    Returns:
        {stable, score, grade, reasons, com, leg_count, surface_count, per_part_status}
    """
    reasons: List[str] = []
    per_part_status: Dict[str, str] = {p["id"]: "neutral" for p in parts}
    score = 100

    if not parts:
        return {
            "stable": False, "score": 0, "grade": "D",
            "reasons": ["No parts placed"],
            "com": {"x": 0.0, "y": 0.0, "z": 0.0},
            "leg_count": 0, "surface_count": 0,
            "per_part_status": {},
        }

    # Categorise parts
    surfaces   = [p for p in parts if _is_surface(PARTS.get(p.get("type",""), {}).get("kind"))]
    legs       = [p for p in parts if PARTS.get(p.get("type",""), {}).get("kind") == "leg"]
    backs      = [p for p in parts if PARTS.get(p.get("type",""), {}).get("kind") == "back"]
    panels_all = [p for p in parts if PARTS.get(p.get("type",""), {}).get("kind") == "panel"]

    # Centre of mass
    total_mass = sum(PARTS.get(p.get("type",""), {}).get("kg", 0) for p in parts)
    com_x = com_y = com_z = 0.0
    if total_mass > 0:
        for p in parts:
            kg = PARTS.get(p.get("type",""), {}).get("kg", 0)
            com_x += p.get("x", 0) * kg
            com_y += p.get("y", 0) * kg
            com_z += p.get("z", 0) * kg
        com_x /= total_mass
        com_y /= total_mass
        com_z /= total_mass

    # Basic checks
    if not surfaces:
        reasons.append("No surface detected")
        score -= 40
    if not legs and not panels_all:
        reasons.append("No legs or panels")
        score -= 40

    TOL = 0.12  # 12 cm tolerance for leg-surface contact

    for surf in surfaces:
        sb = get_bounds(surf)
        if not sb:
            continue
        surf_bottom = sb["min_y"]

        # Find legs supporting this surface
        s_legs = [
            l for l in legs
            if (b := get_bounds(l)) and abs(b["max_y"] - surf_bottom) < TOL
        ]
        # Find panels supporting this surface
        s_panels = [
            p for p in panels_all
            if (b := get_bounds(p)) and b["max_y"] >= surf_bottom - TOL
        ]

        if len(s_legs) >= 3 or len(s_panels) >= 2:
            per_part_status[surf["id"]] = "stable"
            for l in s_legs:
                per_part_status[l["id"]] = "stable"
            for p in s_panels:
                per_part_status[p["id"]] = "stable"

            # Height consistency check
            if len(s_legs) > 1:
                bottoms = [get_bounds(l)["min_y"] for l in s_legs if get_bounds(l)]
                if bottoms and (max(bottoms) - min(bottoms)) > 0.08:
                    reasons.append("Legs have unequal heights")
                    score -= 25
                    per_part_status[surf["id"]] = "unstable"
                    for l in s_legs:
                        per_part_status[l["id"]] = "unstable"

            # CoM within leg polygon check (±5cm tolerance)
            if len(s_legs) >= 2:
                lxs = [l.get("x", 0) for l in s_legs]
                lzs = [l.get("z", 0) for l in s_legs]
                if (
                    com_x < min(lxs) - 0.05 or com_x > max(lxs) + 0.05 or
                    com_z < min(lzs) - 0.05 or com_z > max(lzs) + 0.05
                ):
                    reasons.append("Centre of mass outside support base")
                    score -= 30
                    per_part_status[surf["id"]] = "unstable"
        else:
            per_part_status[surf["id"]] = "unstable"
            if s_legs or s_panels:
                reasons.append("Insufficient support (need 3+ legs or 2+ panels)")
                score -= 20
            else:
                reasons.append("Surface has no support")
                score -= 35

        # Backrest connectivity check
        sx_span = (sb["max_x"] - sb["min_x"]) / 2 + 0.1
        for b in backs:
            in_range = (
                abs(b.get("x", 0) - surf.get("x", 0)) < sx_span and
                abs(b.get("z", 0) - surf.get("z", 0)) < 0.3
            )
            per_part_status[b["id"]] = "stable" if in_range else "unstable"
            if not in_range:
                reasons.append("Backrest disconnected")
                score -= 15

    # Ground-level neutrals become stable
    for p in parts:
        if per_part_status[p["id"]] == "neutral":
            b = get_bounds(p)
            if b and b["min_y"] < 0.04:
                per_part_status[p["id"]] = "stable"

    score = max(0, min(100, score))
    stable = score >= 70 and not reasons
    if stable:
        reasons.append("Structure is stable")

    grade = "A" if score >= 90 else "B" if score >= 75 else "C" if score >= 50 else "D"

    return {
        "stable": stable,
        "score": score,
        "grade": grade,
        "reasons": reasons,
        "com": {"x": round(com_x, 3), "y": round(com_y, 3), "z": round(com_z, 3)},
        "leg_count": len(legs),
        "surface_count": len(surfaces),
        "per_part_status": per_part_status,
    }
